examples/experimental/native/flexible_primitives_mnist_pytorch/model_def.py

```diff
@@ -0,0 +1,199 @@
+        self.download_directory = f"/tmp/data-rank{self.context.distributed.get_rank()}"
+        self.data_downloaded = False
+
+        self.model = self.context._Model(GAN(trial_context))
+
+        lr = self.context.get_hparam("lr")
+        b1 = self.context.get_hparam("b1")
+        b2 = self.context.get_hparam("b2")
+
+        self.opt_g = self.context._Optimizer(torch.optim.Adam(self.model.generator.parameters(), lr=lr, betas=(b1, b2)))
+        self.opt_d = self.context._Optimizer(torch.optim.Adam(self.model.discriminator.parameters(), lr=lr, betas=(b1, b2)))
+
+        # self.model, (self.opt_g, self.opt_d) = self.context._configure_apex_amp(
+        #     model=self.model,
+        #     optimizers=[self.opt_g, self.opt_d],
+        #     opt_level="O1",
+        # )
+
+        self.lr_g = self.context._LRScheduler(
+            lr_scheduler=LambdaLR(self.opt_g, lr_lambda=lambda epoch: 0.95 ** epoch),
+            step_mode=LRScheduler.StepMode.STEP_EVERY_EPOCH,
+        )
+
+    def build_training_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        train_data = data.get_dataset(self.download_directory, train=True)
+        return DataLoader(train_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def build_validation_data_loader(self) -> DataLoader:
+        if not self.data_downloaded:
+            self.download_directory = data.download_dataset(
+                download_directory=self.download_directory,
+                data_config=self.context.get_data_config(),
+            )
+            self.data_downloaded = True
+
+        validation_data = data.get_dataset(self.download_directory, train=False)
+        return DataLoader(validation_data, batch_size=self.context.get_per_slot_batch_size())
+
+    def train_batch(
+        self, batch: TorchData, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+        imgs, _ = batch
+
+        # train generator
+        # sample noise and match gpu device (or keep as cpu)
+        z = torch.randn(imgs.shape[0], self.context.get_hparam("latent_dim"))
+        z = self.context._to_device(z)
+
+        # generate images
+        generated_imgs = self.model.generator(z)
+
+        # log sampled images to current directory
+        sample_imgs = generated_imgs[:6]
+        grid = torchvision.utils.make_grid(sample_imgs)
+        self.logger.writer.add_image(f'generated_images_epoch_{epoch_idx}', grid, batch_idx)
+
+        # ground truth result (ie: all fake)
+        # put on GPU because we created this tensor inside training_loop
+        valid = torch.ones(imgs.size(0), 1)
+        valid = self.context._to_device(valid)
```

Comment:
General question: would a user ever try to do something like this?  Like should handling this be a concern for us?

API question: is it possible for us to handle this in a `forward` hook when the user calls `model.discriminator(their model?  Also: is it a good idea? (off the top of my head, I'm guessing the answers for those two questions are "yes", and "no" but I want to know what you think)

I am also not sure what happens if the user totally forgets to call `_to_device()` entirely, do you know?

Reply:
> would a user ever try to do something like this? Like should handling this be a concern for us?

Yes. A user might generate data, which is used in training. We need to move the data to the device allocated to the training process. In this GAN model, fake data are generated and used to train both the generator and discriminator.

>  is it possible for us to handle this in a forward hook when the user calls `model.discriminator(their model? Also: is it a good idea?

This sounds like monkey patching `forward` of `nn.Module`. Doesn't sound good idea to me.

That throws errors. I actually tried:
```
RuntimeError: Expected object of device type cuda but got device type cpu for argument #2 'mat1' in call to _th_addmm
```

Reply:
It seems unfortunate that users would have to call `to_device()` ever, since we handle it everywhere else, but I suppose if the user needs it they need it.

But... why would they generate data in the middle of their `train_batch()` function?  The example here seems kinda contrived to me.

Reply:
In GAN, users generate a batch of random images for training for each batch of real data. We can definitely let them handle that in data loaders but it is not how people write Pytorch. If they have a GAN pytorch model, typically the code is written in a style that the random images are created on each batch rather than in a random data loader. That users handle `to_device()` doesn't too bad to me because at least in Pytorch and PytorchLightning they need to do it themselves so users who have used GPU with Pytorch would know how to handle.

Reply:
I agree exposing `to_device` is out of the scope of this project and GAN is not what we focus on. But I can't think of a better API.

Reply:
Ok, this makes sense to me.